src/rpc/claimtrie.cpp

```diff
@@ -323,6 +349,42 @@ UniValue getclaimsforname(const UniValue& params, bool fHelp)
 
     LOCK(cs_main);
     std::string name = params[0].get_str();
+
+    UniValue ret(UniValue::VOBJ);
+
+    std::vector<CBlockIndex*> indexes;
+    CCoinsViewCache coins(pcoinsTip);
+    CClaimTrieCache claims(pclaimTrie);
+    CValidationState state;
+
+    BOOST_SCOPE_EXIT(&indexes, &coins, &claims, &state) {
+        // restore the rolled-back state
+        for (std::size_t i = indexes.size(); i > 0; --i) {
+            CBlock block;
+            ReadBlockFromDisk(block, indexes.back(), Params().GetConsensus());
+            ConnectBlock(block, state, indexes[i - 1], coins, claims);
+        }
+        if (!indexes.empty()) {
+            // necessary ?
+            coins.Flush();
+            claims.flush();
+        }
+    } BOOST_SCOPE_EXIT_END
+
+    if (params.size() > 1)
+    {
+        int minconf = params[1].get_int();
+        while(--minconf > 0) {
+            indexes.push_back(chainActive.Tip());
+            CBlock block;
+            ReadBlockFromDisk(block, indexes.back(), Params().GetConsensus());
+            DisconnectBlock(block, state, indexes.back(), coins, claims);
```

Comment:
In this method we are actually rolling back the real blocks -- not just the caches. We hold the locks and disconnect some number of tips and then rely on our scope-exit handler to put everything back before we leave. I think it's a horrible approach, really, but didn't see a good way to do it otherwise in the current codebase. We lock the entire system for handling this method. That's so wrong. I think Roy's approach of storing all claimtrie (nodes that have ever existed) in LevelDB is much better -- the old nodes would still be there and you could just start a search at a certain time period -- we wouldn't ever have to lock; we could run numerous calls concurrently.

Reply:
All (current) claimtrie nodes are stored in leveldb.  Storing all changes over time (while nice for analysis) isn't actually that practical of a feature to have.  Eeven concurrent performance on lookups isn't needed for just about anything practical (other than analysis), where I'd argue a relational DB is just as well suited, if not better.  Horrible or not, all persistent state changes of a blockchain occur via connecting and disconnecting blocks.

Now is all this needed for this particular issue?  I'm not sure yet, haven't thought about it enough.